File: GamePhysicsSdk/EasyPhisicsDotNet/EpxBroadPhase.cs
using System;
using System.Collections.Generic;

namespace EasyPhisicsDotNet
{
    public class EpxBroadPhase
    {
        static readonly EpxVector3 AabbExpand = new EpxVector3(0.01f);

        /// <summary>
        /// ブロードフェーズ
        /// </summary>
        /// <param name="states">剛体の状態の配列</param>
        /// <param name="collidables">剛体の形状の配列</param>
        /// <param name="oldPairs">前フレームのペア配列</param>
        /// <param name="callback">コールバック</param>
        /// <returns>検出されたペア配列</returns>
        public static List<EpxPair> Execute(
            List<EpxState> states,
            List<EpxCollidable> collidables,
            List<EpxPair> oldPairs,
            IEpxBroadPhaseCallback callback)
        {
            // ペアを検出する
            var newPairs = _CreatePairs(states, collidables, callback);

            // 検出されたペアのうち、既存のものについては、衝突情報を更新の上、再利用する。
            _ReuseOldContactInformation(newPairs, oldPairs, states);

            return newPairs;
        }

        /// <summary>
        /// AABB交差ペアを見つける（総当たり）
        /// 処理の内容を明確にするため、ここでは空間分割テクニックを使っていませんが、
        /// 理論編で解説されているSweet and prune等の手法を使えば高速化できます。
        /// </summary>
        /// <param name="states"></param>
        /// <param name="collidables"></param>
        /// <param name="callback"></param>
        /// <returns></returns>
        static List<EpxPair> _CreatePairs(List<EpxState> states, List<EpxCollidable> collidables, IEpxBroadPhaseCallback callback)
        {
            var pairs = new List<EpxPair>();
            for (int i = 0; i < states.Count; i++)
            {
                EpxVector3 centerA; // 中心座標
                EpxVector3 halfA;   // 大きさ
                _GetAabb(states[i], collidables[i], out centerA, out halfA);
                for (int j = i + 1; j < states.Count; j++)
                {
                    if (callback != null)
                        if (!callback.Execute(i, j))
                            continue;
                    EpxVector3 centerB;// 中心座標
                    EpxVector3 halfB; // 大きさ
                    _GetAabb(states[j], collidables[j], out centerB, out halfB);
                    if (_EpxIntersectAABB(centerA, halfA, centerB, halfB))
                        pairs.Add(new EpxPair(i, j));
                }
            }
            return pairs;
        }

        static void _GetAabb(EpxState state, EpxCollidable collidable, out EpxVector3 center, out EpxVector3 half)
        {
            center = state.Transform.Transform(collidable.Center);
            half = _AbsPerElem(state.Transform.Rotate(collidable.Half + AabbExpand));// AABBサイズを若干拡張
        }

        static EpxVector3 _AbsPerElem(EpxVector3 v)
        {
            return new EpxVector3((float)Math.Abs(v.X), (float)Math.Abs(v.Y), (float)Math.Abs(v.Z));
        }

        static bool _EpxIntersectAABB(EpxVector3 centerA, EpxVector3 halfA, EpxVector3 centerB, EpxVector3 halfB)
        {
            if (Math.Abs(centerA[0] - centerB[0]) > halfA[0] + halfB[0])
                return false;
            if (Math.Abs(centerA[1] - centerB[1]) > halfA[1] + halfB[1])
                return false;
            if (Math.Abs(centerA[2] - centerB[2]) > halfA[2] + halfB[2])
                return false;
            return true;
        }

        static void _ReuseOldContactInformation(
            List<EpxPair> newPairs,
            List<EpxPair> oldPairs,
            List<EpxState> states)
        {
            if (oldPairs.Count == 0)
                return;
            var ope = oldPairs.GetEnumerator();
            foreach (EpxPair n in newPairs)
            {
                var o = ope.Current;
                if (n.Key < o.Key)
                    continue;
                if (n.Key == o.Key)
                {
                    // ブロードフェイズで衝突情報を更新するのではなく、ナローフェイズで更新するほうがわかりやすいのではないか?(秋山)
                    EpxContact c = o.Contact;
                    c.Refresh(
                        states[o.rigidBodyA].Transform,
                        states[o.rigidBodyB].Transform);
                    n.Contact = c;
                }
                if (!ope.MoveNext())
                    return;
            }
        }
    }
}

File: GamePhysicsSdk/easy_physics/pipeline/EpxBroadphase.cpp
/*
	Copyright (c) 2012 Hiroshi Matsuike

	This software is provided 'as-is', without any express or implied
	warranty. In no event will the authors be held liable for any damages
	arising from the use of this software.

	Permission is granted to anyone to use this software for any purpose,
	including commercial applications, and to alter it and redistribute it
	freely, subject to the following restrictions:

	1. The origin of this software must not be misrepresented; you must not
	claim that you wrote the original software. If you use this software
	in a product, an acknowledgment in the product documentation would be
	appreciated but is not required.

	2. Altered source versions must be plainly marked as such, and must not be
	misrepresented as being the original software.

	3. This notice may not be removed or altered from any source distribution.
*/

#include "EpxBroadphase.h"
#include "EpxSort.h"

#define EPX_AABB_EXPAND 0.01f

namespace EasyPhysics {

static inline 
EpxBool epxIntersectAABB(const EpxVector3 &centerA,const EpxVector3 &halfA,const EpxVector3 &centerB,const EpxVector3 &halfB)
{
	if(fabs(centerA[0] - centerB[0]) > halfA[0] + halfB[0]) return false;
	if(fabs(centerA[1] - centerB[1]) > halfA[1] + halfB[1]) return false;
	if(fabs(centerA[2] - centerB[2]) > halfA[2] + halfB[2]) return false;
	return true;
}

void epxBroadPhase(
	const EpxState *states,
	const EpxCollidable *collidables,
	EpxUInt32 numRigidBodies,
	const EpxPair *oldPairs,
	const EpxUInt32 numOldPairs,
	EpxPair *newPairs,
	EpxUInt32 &numNewPairs,
	const EpxUInt32 maxPairs,
	EpxAllocator *allocator,
	void *userData,
	epxBroadPhaseCallback callback)
{
	assert(states);
	assert(collidables);
	assert(oldPairs);
	assert(newPairs);
	assert(allocator);
	
	numNewPairs = 0;
	
	// AABB交差ペアを見つける（総当たり）
	// 処理の内容を明確にするため、ここでは空間分割テクニックを使っていませんが、
	// 理論編で解説されているSweet and prune等の手法を使えば高速化できます。
	for(EpxUInt32 i=0;i<numRigidBodies;i++) {
		for(EpxUInt32 j=i+1;j<numRigidBodies;j++) {
			const EpxState &stateA = states[i];
			const EpxCollidable &collidableA = collidables[i];
			const EpxState &stateB = states[j];
			const EpxCollidable &collidableB = collidables[j];

			if(callback && !callback(i,j,userData)) {
				continue;
			}
			
			EpxMatrix3 orientationA(stateA.m_orientation);
			EpxVector3 centerA = stateA.m_position + orientationA * collidableA.m_center;
			EpxVector3 halfA = absPerElem(orientationA) * ( collidableA.m_half + EpxVector3(EPX_AABB_EXPAND) );// AABBサイズを若干拡張
			
			EpxMatrix3 orientationB(stateB.m_orientation);
			EpxVector3 centerB = stateB.m_position + orientationB * collidableB.m_center;
			EpxVector3 halfB = absPerElem(orientationB) * ( collidableB.m_half + EpxVector3(EPX_AABB_EXPAND) );// AABBサイズを若干拡張
			
			if(epxIntersectAABB(centerA,halfA,centerB,halfB) && numNewPairs < maxPairs) {
				EpxPair &newPair = newPairs[numNewPairs++];
				
				newPair.rigidBodyA = i<j?i:j;
				newPair.rigidBodyB = i<j?j:i;
				newPair.contact = NULL;
			}
		}
	}
	
	// ソート
	{
		EpxPair *sortBuff = (EpxPair*)allocator->allocate(sizeof(EpxPair)*numNewPairs);
		epxSort<EpxPair>(newPairs,sortBuff,numNewPairs);
		allocator->deallocate(sortBuff);
	}
	
	// 新しく検出したペアと過去のペアを比較
	EpxPair *outNewPairs = (EpxPair*)allocator->allocate(sizeof(EpxPair)*numNewPairs);
	EpxPair *outKeepPairs = (EpxPair*)allocator->allocate(sizeof(EpxPair)*numOldPairs);
	assert(outNewPairs);
	assert(outKeepPairs);
	
	EpxUInt32 nNew = 0;
	EpxUInt32 nKeep = 0;
	
	EpxUInt32 oldId = 0,newId = 0;
	
	while(oldId<numOldPairs&&newId<numNewPairs) {
		if(newPairs[newId].key > oldPairs[oldId].key) {
			// remove
			allocator->deallocate(oldPairs[oldId].contact);
			oldId++;
		}
		else if(newPairs[newId].key == oldPairs[oldId].key) {
			// keep
			assert(nKeep<=numOldPairs);
			outKeepPairs[nKeep] = oldPairs[oldId];
			nKeep++;
			oldId++;
			newId++;
		}
		else {
			// new
			assert(nNew<=numNewPairs);
			outNewPairs[nNew] = newPairs[newId];
			nNew++;
			newId++;
		}
	};
	
	if(newId<numNewPairs) {
		// all new
		for(;newId<numNewPairs;newId++,nNew++) {
			assert(nNew<=numNewPairs);
			outNewPairs[nNew] = newPairs[newId];
		}
	}
	else if(oldId<numOldPairs) {
		// all remove
		for(;oldId<numOldPairs;oldId++) {
			allocator->deallocate(oldPairs[oldId].contact);
		}
	}
	
	for(EpxUInt32 i=0;i<nNew;i++) {
		outNewPairs[i].contact = (EpxContact*)allocator->allocate(sizeof(EpxContact));
		outNewPairs[i].contact->reset();
	}
	
	for(EpxUInt32 i=0;i<nKeep;i++) {
		outKeepPairs[i].contact->refresh(
			states[outKeepPairs[i].rigidBodyA].m_position,
			states[outKeepPairs[i].rigidBodyA].m_orientation,
			states[outKeepPairs[i].rigidBodyB].m_position,
			states[outKeepPairs[i].rigidBodyB].m_orientation);
	}
	
	numNewPairs = 0;
	for(EpxUInt32 i=0;i<nKeep;i++) {
		outKeepPairs[i].type = EpxPairTypeKeep;
		newPairs[numNewPairs++] = outKeepPairs[i];
	}
	for(EpxUInt32 i=0;i<nNew;i++) {
		outNewPairs[i].type = EpxPairTypeNew;
		newPairs[numNewPairs++] = outNewPairs[i];
	}
	
	allocator->deallocate(outKeepPairs);
	allocator->deallocate(outNewPairs);
	
	// ソート
	{
		EpxPair *sortBuff = (EpxPair*)allocator->allocate(sizeof(EpxPair)*numNewPairs);
		epxSort<EpxPair>(newPairs,sortBuff,numNewPairs);
		allocator->deallocate(sortBuff);
	}
}

} // namespace EasyPhysics
